File: Data Structures & Algorithms/decode-ways/submission-6.cpp
class Solution {
public:
    int dfs(string s, int i, vector<int>& dp) {
        if (dp[i] != -1) return dp[i];
        if (s[i] == '0') return 0;

        int res = dfs(s, i + 1, dp);
        if (i + 1 < s.size() and ((s[i] == '1') or (s[i] == '2' and s[i + 1] >= '0' and s[i + 1] <= '6'))) {
            res += dfs(s, i + 2, dp);
        }

        dp[i] = res;
        return res;
    }

    int numDecodings(string s) {
        vector<int> dp(s.size() + 1, -1);
        dp[s.size()] = 1;
        return dfs(s, 0, dp);
    }
};

File: Data Structures & Algorithms/ones-and-zeroes/submission-6.py
class Solution:
    def find01(self, s):
        zero, one = 0, 0
        for c in s:
            if c == "0": zero += 1
            else: one += 1
        
        return zero, one

    def findMaxForm(self, strs: List[str], m: int, n: int) -> int:
        dp = {} # dp[(i, m, n)] = max len of valid subset starting there
        def dfs(i, fake_m, fake_n):
            if i == len(strs):
                return 0
            
            if (i, fake_m, fake_n) in dp:
                return dp[(i, fake_m, fake_n)]
            
            zero, one = self.find01(strs[i])
            
            take = 0
            if fake_m >= zero and fake_n >= one:
                take = 1 + dfs(i + 1, fake_m - zero, fake_n - one)

            skip = dfs(i + 1, fake_m, fake_n)
            res = max(take, skip)
            dp[(i, fake_m, fake_n)] = res
            return res

        return dfs(0, m, n)
